okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimitingInterceptor.java

```diff
@@ -0,0 +1,163 @@
+ * will be returned to the pool before acquiring a new one.
+ *
+ * Users must also wrap the final callback they use; this is used in two ways; first it clears the state in case of
+ * failure, secondly on success it will wait until the response is closed before handing back permits. In other words,
+ * if you have a server with a concurrency limit (e.g. it is CPU bound), clients should respect the server's
+ * concurrency limit.
+ *
+ * This has a timeout of 1 minute (before an error is logged) in order to try to catch people who have leaked responses
+ * (which here will deadlock otherwise). It indicates an application bug every time, but might affect users poorly.
+ * I'm happy to remove it, but think there should probably be another solution?
+ */
+final class ConcurrencyLimitingInterceptor implements Interceptor {
+    private static final Logger log = LoggerFactory.getLogger(ConcurrencyLimitingInterceptor.class);
+    private final ConcurrencyLimiters limiters = new ConcurrencyLimiters();
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        ConcurrencyLimitTag tagState = chain.request().tag(ConcurrencyLimitTag.class);
+        tagState.invalidate();
+        tagState.setListener(limiters.limiter(chain.request()));
+        return chain.proceed(chain.request());
+    }
+
+    public static Callback wrapCallback(Callback callback) {
+        return new Callback() {
+            @Override
+            public void onFailure(Call call, IOException e) {
+                Optional.ofNullable(call.request().tag(ConcurrencyLimitTag.class)).ifPresent(ConcurrencyLimitTag::invalidate);
+                callback.onFailure(call, e);
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                Response newResponse =
+                        Optional.ofNullable(call.request().tag(ConcurrencyLimitTag.class))
+                                .map(t -> wrapResponse(t, response))
+                        .orElse(response);
+                callback.onResponse(call, newResponse);
+            }
+        };
+    }
+
+    public static Request wrapRequest(Request request) {
+        return request.newBuilder().tag(ConcurrencyLimitTag.class, new ConcurrencyLimitTag()).build();
+    }
+
+    private static Response wrapResponse(ConcurrencyLimitTag tag, Response response) {
+        if (response.body() == null) {
+            return response;
+        }
+        ResponseBody currentBody = response.body();
+        ResourceDeallocator deallocator = new ResourceDeallocator(tag);
+        ResponseBody newResponseBody =
+                ResponseBody.create(currentBody.contentType(), currentBody.contentLength(),
+                        new ReleaseConcurrencyLimitBufferedSource(currentBody.source(), tag, deallocator));
+        deallocator.timeout(1, TimeUnit.MINUTES);
```

Comment:
How does this play with long streaming requests? Should I release the limiter as soon as I start streaming the data, rather than after all streaming is done?

Reply:
At present the assumption is that on the server side, if you have particularly cheap requests the server should be slackening its guarantees. For this reason, I'd prefer to keep the criteria here 'is the request in flight'? because then it can match up with whatever the server is doing.

Reply:
clarified in person: in the new code the timeout only applies to the period when you are waiting to receive a limiter token; the limiter for the long-running streaming request will not time out when the request is being served.

If new requests come in and spend over a minute waiting to get a rate limiting token (because we're at max capacity for a really long time), we will circuit-break and effectively reset the rate limiting, with the assumption that this was a resource leak and not actually-at-full-capacity-unable-to-serve-a-single-request for over a minute.